File: bytopic/backtracking/java/group3/NQueensII.java
// The n-queens puzzle is the problem of placing n queens on an n x n chessboard such that
// no two queens attack each other.
// Given an integer n, return the number of distinct solutions to the n-queens puzzle.
// Example:
// Input: n = 4
// Output: 2
// Explanation: There are two distinct solutions to the 4-queens puzzle as shown.
// Note: Find visuals to understand
//
// Tag: 105/150
// Tag: 52/2927, R1500/2936 (overall frequency ranking)

import java.util.Arrays;
import java.util.ArrayList;
import java.util.List;

class NQueensII {
    public static void main(String...args) {
        int n = 4;
        System.out.println("Num. of distinct solutions: " + totalNQueens(n));
    }

    static int totalNQueens(int n) {
        char board[][] = new char[n][n];
        for(char[] c : board)
            Arrays.fill(c, '.');
        return dfs(0, board);
    }

    private static int dfs(int col, char board[][]) {
        if(col == board.length) return 1;
        int count = 0;
        for(int row = 0; row < board.length; row++){
            if(isSafe(board, row, col)){
                board[row][col] = 'Q';
                count += dfs(col + 1, board);
                board[row][col] = '.';
            }
        }
        return count;
    }

    private static boolean isSafe(char board[][], int row, int col) {
        int dupRow = row;
        int dupCol = col;

        while(row >= 0 && col >= 0){
            if(board[row][col] == 'Q') return false;
            row--;
            col--;
        }

        row = dupRow;
        col = dupCol;
        while(col >= 0){
            if(board[row][col] == 'Q') return false;
            col--;
        }

        row = dupRow;
        col = dupCol;
        while(col >= 0 && row < board.length){
            if(board[row][col] == 'Q') return false;
            row++;
            col--;
        }
        return true;
    }
}

File: bytopic/backtracking/python/group3/n_queens_ii.py

"""
The n-queens puzzle is the problem of placing n queens on an n x n chessboard such that
no two queens attack each other.
Given an integer n, return the number of distinct solutions to the n-queens puzzle.

Example:
Input: n = 4
Output: 2
Explanation: There are two distinct solutions to the 4-queens puzzle as shown.
Note: Find visuals to understand

Tag: 105/150
Tag: 52/2927, R1500/2936 (overall frequency ranking)
"""


def total_n_queens(n: int) -> int:
    visited_cols = set()
    visited_diagonals = set()
    visited_anti_diagonals = set()

    def backtrack(r):
        if r == n:  # Valid solution state
            return 1

        cnt = 0
        for c in range(n):
            if not (c in visited_cols or (r-c) in visited_diagonals or (r+c) in visited_anti_diagonals):
                visited_cols.add(c)
                visited_diagonals.add(r - c)
                visited_anti_diagonals.add(r + c)
                cnt += backtrack(r + 1)  # count the overall tally from this current state

                visited_cols.remove(c)
                visited_diagonals.remove(r - c)
                visited_anti_diagonals.remove(r + c)
        return cnt
    return backtrack(0)


def main():
    n = 4
    print(total_n_queens(n))


if __name__ == "__main__":
    main()
